### PR TITLE
feature/1365_data_capping_mysql

### PR DESCRIPTION
**Pull request by frbattid**
* Implements issue #1365 (capping part)
* PR https://github.com/telefonicaid/fiware-cygnus/pull/1360 must be merged before this one.
* (Unofficial) e2e tests passed:
```
mysql> select * from traffic_Room1_Room order by recvTime;
+---------------+-------------------------+-------------------+----------+------------+-------------+------------+-----------+--------+
| recvTimeTs    | recvTime                | fiwareServicePath | entityId | entityType | attrName    | attrType   | attrValue | attrMd |
+---------------+-------------------------+-------------------+----------+------------+-------------+------------+-----------+--------+
| 1483021054473 | 2016-12-29T14:17:34.473 | /traffic          | Room1    | Room       | temperature | centigrade | 26.5      | []     |
| 1483021120977 | 2016-12-29T14:18:40.977 | /traffic          | Room1    | Room       | temperature | centigrade | 26.5      | []     |
| 1483021121407 | 2016-12-29T14:18:41.407 | /traffic          | Room1    | Room       | temperature | centigrade | 26.5      | []     |
| 1483021518722 | 2016-12-29T14:25:18.722 | /traffic          | Room1    | Room       | temperature | centigrade | 26.5      | []     |
| 1483021531954 | 2016-12-29T14:25:31.954 | /traffic          | Room1    | Room       | temperature | centigrade | 26.5      | []     |
+---------------+-------------------------+-------------------+----------+------------+-------------+------------+-----------+--------+
5 rows in set (0.00 sec)
..
..
time=2016-12-29T14:26:05.270UTC | lvl=DEBUG | corr=59f5554d-20fe-428b-b78c-07cf556decfc | trans=59f5554d-20fe-428b-b78c-07cf556decfc | srv=sc_madrid | subsrv=/traffic | comp=cygnus-ngsi | op=delete | msg=com.telefonica.iot.cygnus.backends.mysql.MySQLBackendImpl[200] : Executing MySQL query 'delete from `traffic_Room1_Room` where recvTime='2016-12-29T14:17:34.473' or recvTime='2016-12-29T14:18:40.977' or recvTime='2016-12-29T14:18:41.407''
..
..
mysql> select * from traffic_Room1_Room order by recvTime;
+---------------+-------------------------+-------------------+----------+------------+-------------+------------+-----------+--------+
| recvTimeTs    | recvTime                | fiwareServicePath | entityId | entityType | attrName    | attrType   | attrValue | attrMd |
+---------------+-------------------------+-------------------+----------+------------+-------------+------------+-----------+--------+
| 1483021518722 | 2016-12-29T14:25:18.722 | /traffic          | Room1    | Room       | temperature | centigrade | 26.5      | []     |
| 1483021531954 | 2016-12-29T14:25:31.954 | /traffic          | Room1    | Room       | temperature | centigrade | 26.5      | []     |
| 1483021556434 | 2016-12-29T14:25:56.434 | /traffic          | Room1    | Room       | temperature | centigrade | 26.5      | []     |
| 1483021557044 | 2016-12-29T14:25:57.44  | /traffic          | Room1    | Room       | temperature | centigrade | 26.5      | []     |
| 1483021557529 | 2016-12-29T14:25:57.529 | /traffic          | Room1    | Room       | temperature | centigrade | 26.5      | []     |
+---------------+-------------------------+-------------------+----------+------------+-------------+------------+-----------+--------+
5 rows in set (0.00 sec)
```

### CODE REVIEW

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md [hunk 1]
```diff
@@ -214,6 +215,9 @@ If `attr_persistence=colum` then `NGSIMySQLSink` will persist the data within th
 | batch\_timeout | no | 30 | Number of seconds the batch will be building before it is persisted as it is. |
 | batch\_ttl | no | 10 | Number of retries when a batch cannot be persisted. Use `0` for no retries, `-1` for infinite retries. Please, consider an infinite TTL (even a very large one) may consume all the sink's channel capacity very quickly. |
 | batch\_retry\_intervals | no | 5000 | Comma-separated list of intervals (in miliseconds) at which the retries regarding not persisted batches will be done. First retry will be done as many miliseconds after as the first value, then the second retry will be done as many miliseconds after as second value, and so on. If the batch\_ttl is greater than the number of intervals, the last interval is repeated. |
+| persistence\_policy.max_records | no | -1 | Maximum number of records allowed for a table before it is capped. `-1` disables this policy. |
+| persistence\_policy.expiration_time | no | -1 | Maximum number of seconds a record is maintained in a table before expiration. `-1` disables this policy. |
+| persistence\_policy.checking_time | no | 3600 | Frequency (in seconds) at which the sink ckecks for record expiration. |
```
Comment (fgalan) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md]: Typo: ckecks

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md [hunk 2]
```diff
@@ -301,6 +308,14 @@ Despite the old encoding will be deprecated in the future, it is possible to swi
 
 [Top](#top)
 
+####<a name="section2.3.6"></a>About capping resources and expirating records
+Capping and expiration are disabled by default. Nevertheless, if desired, this can be enabled:
+
+* Capping by the number of records. This allows the resource growing up until certain configured maximum number of records is reached (`persistence_policy.max_records`), and then maintains a such a constant number of records.
```
Comment (fgalan) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md]: "a such a constant ('a' twice)"... sounds a bit weird. Maybe it is a typo.

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md [hunk 3]
```diff
@@ -301,6 +308,14 @@ Despite the old encoding will be deprecated in the future, it is possible to swi
...
+* Expirating by time the records. This allows the resource growing up until records become old, i.e. overcome certain configured expiration time (`persistence_policy.expiration_time`).
```
Comment (fgalan) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md]: overcome -> overpass ?

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md [hunk 4]
```diff
@@ -309,6 +324,14 @@ As any other NGSI-like sink, `NGSIMySQLSink` extends the base `NGSISink`. The me
 
 A `Batch` contains a set of `NGSIEvent` objects, which are the result of parsing the notified context data events. Data within the batch is classified by destination, and in the end, a destination specifies the MySQL table where the data is going to be persisted. Thus, each destination is iterated in order to compose a per-destination data string to be persisted thanks to any `MySQLBackend` implementation.
 
+    void capRecords(NGSIBatch batch, long maxRecords) throws EventDeliveryException;
+    
+This method is always called immediatelly after `persistBacth()`. The same destination tables that were upserted are now checked in terms of number of records: if the configured maximum (`persistence_policy.max_records`) is overcome for any of the updated tables, then as many oldest records are deleted as required until the maximum number of records is reached.
```
Comment (fgalan) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md]: overcome -> overpassed ?

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md [hunk 5]
```diff
@@ -309,6 +324,14 @@ As any other NGSI-like sink, `NGSIMySQLSink` extends the base `NGSISink`. The me
...
+    
+    void expirateRecords(long expirationTime);
+    
+This method is called in a periodical way (based on `persistence_policy.checking_time`), and if the configured expiration time (`persistence_policy.expiration_time`) is overcome for any of the records within any of the tables, then it is deleted.
```
Comment (fgalan) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md]: overcome -> overpassed?

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md (hunk 1 above)
Comment (frbattid) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1366/commits/536811ecc9e7c7382eac81509a89ad3a34121fdf

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md (hunk 2 above)
Comment (frbattid) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1366/commits/536811ecc9e7c7382eac81509a89ad3a34121fdf

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md (hunk 3 above)
Comment (frbattid) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1366/commits/536811ecc9e7c7382eac81509a89ad3a34121fdf

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md (hunk 4 above)
Comment (frbattid) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1366/commits/536811ecc9e7c7382eac81509a89ad3a34121fdf

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md (hunk 5 above)
Comment (frbattid) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_mysql_sink.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1366/commits/536811ecc9e7c7382eac81509a89ad3a34121fdf

File: cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mysql/MySQLBackendImpl.java [hunk 6]
```diff
@@ -147,30 +151,131 @@ public void insertContextData(String dbName, String tableName, String fieldNames
...
+        Connection con = driver.getConnection(dbName);
+            
+        try {
+            stmt = con.createStatement();
+        } catch (SQLException e) {
+            throw new CygnusRuntimeError("SQLException, " + e.getMessage());
+        } // try catch
+        
+        try {
+            String query = "select " + selection + " from `" + tableName + "` order by recvTime";
```
Comment (fgalan) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mysql/MySQLBackendImpl.java]: Maybe at this point a FIXME/TODO comment pointing to https://github.com/telefonicaid/fiware-cygnus/issues/1371 should be added

Comment (frbattid) [inline cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/mysql/MySQLBackendImpl.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1366/commits/b76393a614e92f38e1773a4b467743184cbcd0d6

**Comment by fgalan** [comment]
LGTM

**Pull request merged**